rust/agents/relayer/src/msg/gas_payment/policies/on_chain_fee_quoting.rs

```diff
@@ -0,0 +1,193 @@
+use async_trait::async_trait;
+use eyre::Result;
+
+use hyperlane_core::{
+    HyperlaneMessage, InterchainGasExpenditure, InterchainGasPayment, TxCostEstimate, U256,
+};
+
+use crate::msg::gas_payment::GasPaymentPolicy;
+
+#[derive(Debug)]
+pub struct GasPaymentPolicyOnChainFeeQuoting {
+    /// Numerator value to modify the estimated gas by. The estimated gas value
+    /// is multiplied by this value.
+    fractional_numerator: u64,
+    /// Denominator value to modify the estimated gas by. The estimated gas
+    /// value is divided by this value.
+    fractional_denominator: u64,
+}
+
+impl GasPaymentPolicyOnChainFeeQuoting {
+    pub fn new(fractional_numerator: u64, fractional_denominator: u64) -> Self {
+        Self {
+            fractional_numerator,
+            fractional_denominator,
+        }
+    }
+}
+
+impl Default for GasPaymentPolicyOnChainFeeQuoting {
+    fn default() -> Self {
+        // default to requiring they have paid 1/2 the estimated gas.
+        Self {
+            fractional_numerator: 1,
+            fractional_denominator: 2,
+        }
+    }
+}
+
+#[async_trait]
+impl GasPaymentPolicy for GasPaymentPolicyOnChainFeeQuoting {
+    async fn message_meets_gas_payment_requirement(
+        &self,
+        _message: &HyperlaneMessage,
+        current_payment: &InterchainGasPayment,
+        current_expenditure: &InterchainGasExpenditure,
+        tx_cost_estimate: &TxCostEstimate,
+    ) -> Result<Option<U256>> {
+        let fractional_gas_estimate =
+            tx_cost_estimate.gas_limit * self.fractional_numerator / self.fractional_denominator;
+        let gas_amount = current_payment.gas_amount - current_expenditure.gas_used;
```

Comment:
wonder if this should be a `saturating_sub` just to be safe

Reply:
Good catch, probably would have paniced occurred eventually.